### PR TITLE
Update libMesh

### PR DESCRIPTION
**Pull request by lindsayad**
Summary of changes:  

  * Avoid raw new, use libmesh_make_unique instead
  * Enable PETSc shell matrices for SLEPc eigenvalue solver 
  * DynaIO DoF constraints
  * Barebones-build fixes
  * Specify override to avoid clang compiler warning
  * Add missing PYRAMID cases to l2_lagrange_n_dofs()
  * Nemesis localize fix
  * Sync nodes after all_second_order if not replicated
  * Add a couple DistributedMesh ghosted_elem APIs
  * Update in RB code
  *  Add BoundaryInfo::erase_if() helper function
  * Add BoundaryInfo::remove_node(node, id).
  * Add configure test for std::thread::hardware_concurrency()
  * Copy point locator tolerance in MeshFunction::clone().
  * Add a missing header inclusion.
  * Fixes for --enable-petsc --disable-mpi configurations
  * Add get_row method to Sparse Matrix
  * Allow adaptivity_ex3 to run with more variables
  * RBConstruction: Add post processing callbacks.
  * Fix a bug on allocating segmentlist in Triangle interface
  * Fix adjoints ex1 QoIs for multithreaded runs
  * Threads fixes
  * Add nonblocking packed range push
  * Use DirichletBoundary in example
  * TIMPI configuration tweaks
  * Move non-libMesh-specific Parallel:: code to a subpackage
  * Sparsity pattern optimization for "spider" nodes
  * Curl fix
  * Cpflags fix
  * Move forward declaration into its proper namespace
  * Use ReplicatedMesh in new edgeset unit test
  * Adding "tellp" to ostream_proxy class
  * ignore_warnings.h before tetgen.h
  * DynaIO fixes, calculate_map fix/optimization
  * AbaqusIO: Add helper functions
  * Add support for reading ABAQUS .inp file with *Surface referring to element sets

### CODE REVIEW

**Comment by loganharbour** [comment]
Note that the shift to TIMPI in libMesh caused some issues with _some_ Xcode headers. And by some, I mean it's caused by a specific Xcode install process that wasn't documented. The fix was uninstalling Xcode and CLT, and then re-installing Xcode and CLT.

**Comment by roystgnr** [comment]
In https://github.com/idaholab/moose/pull/14415#issuecomment-557641843 @lindsayad noted an application CI failure due to the move of libMesh::Parallel::StandardType -> TIMPI::StandardType; was this another forward declaration failure (like the one I have that ugly partial-workaround for with Communicator) or was there a new type of breakage?

**Comment by lindsayad** [comment]
I also observed these Mac test failures over on #14167. I cannot reproduce them on Mojave, so I'm wondering whether this is specific to Catalina. I'm probably going to update my laptop OS today, so maybe I can reproduce then

**Comment by permcody** [comment]
I have something to add on these Mac failures. What's happening is a breakdown of the exception handling mechanism. We are throwing a type `Foo` from a downstream stack somewhere and we are not able to catch `Foo` during the unwind process. I have a stand-alone simple C++ code where we were able to demonstrate this. I don't know why it's happening, but we were able to fix it by playing with the plethora of flags that get passed to the compiler for any non-trivial C++ code (i.e. libMesh, MOOSE). In this case we found that the breakage seemed to occur more often when the "-flat-namespace" flag was used. Also, we were able to fix it by changing out one MPI library for another (i.e. going from MPICH to OpenMPI). It's not clear though what's specifically special about the Catalina machine. We started to see these issues when @milljm was working with distributing the stack through Conda instead of a ground-up package builder build. Perhaps some of the compiler flags we use with libMesh MOOSE have deeper implications than we fully understand, but that's essentially what's happening here.

Options: 
 - Dig into compiler, linker hell and figure out what's going on with odd combinations of libraries, compilers that cause exception handling + unwinding (the unwinding part is important here as we can't get simple throw/catch demonstrations to break).
 - Change this one test by adding yet another catch statement (i.e. `catch (...)` that seems to always work even if the specific catch statement fails to catch the exception. The issue with the dot dot dot syntax is that you can't recover specific information about the exception. Do we really care though?

**Comment by permcody** [comment]
While we sort out what we should do to fix the try/catch crap. Maybe we can push the hash forward a bit more? I wouldn't mind picking up the latest Nemesis stuff to start playing with:
https://github.com/libMesh/libmesh/pull/2338

**Comment by lindsayad** [comment]
>     * Change this one test by adding yet another catch statement (i.e. `catch (...)` that seems to always work even if the specific catch statement fails to catch the exception. The issue with the dot dot dot syntax is that you can't recover specific information about the exception. Do we really care though?

That resolves the moose test failure. It looks like there's only one module failure. Unit testing though has a ton of failures.

**Comment by lindsayad** [comment]
It looks like this has affected at least one other project: https://discourse.mc-stan.org/t/dealing-with-catalina/11285

**Comment by permcody** [comment]
@loganharbour said he's not running into this issue on his Catalina machine. Perhaps you just need to know the right number of uninstalls and reinstalls of XCode and CLI-Tools one needs to perform.

**Comment by lindsayad** [comment]
What version of xcode does the Mac test box have? Is it the same as @loganharbour ?

**Comment by lindsayad** [comment]
As @roystgnr suggested this has to be some weird header thing though right? Because this code path has been passing on the Catalina test box before this attempted libmesh update.

**Comment by permcody** [comment]
> As @roystgnr suggested this has to be some weird header thing though right? 

True, but that only adds to the mystery of this whole thing since that would mean this failure is inconsistent (repeatable for a given hash). My stand-alone code is pretty small and fails although I didn't try it on the test machine. I tried it on Jason's machine, which should match the test machine. What a nasty little bug! What's worse is that it may not be a bug at all. We might actually just be doing something wrong somewhere in the build and it's manifesting itself as this odd exception catching failure.

**Comment by lindsayad** [comment]
What do you mean by inconsistent? I would guess this has been repeatable on the test box ever since the header changes in libMesh.

**Comment by roystgnr** [comment]
Specifically what is the class of the exception being thrown and then missed?

**Comment by roystgnr** [comment]
I ask because of the suspicion: is it possible that MOOSE is trying to catch a libmesh exception class but now TIMPI is just throwing a std:: class?  Except in that case I don't see why it wouldn't fail on Linux too.

**Comment by lindsayad** [comment]
The moose test error is a std::domain_error. The unit tests failure look like a mixed bag of std exceptions. So doesn’t seem to be a libmesh to TIMPI exception class issue.

**Comment by permcody** [comment]
Yes, this is where the exception is caught (not caught): 
https://github.com/idaholab/moose/blob/next/test/src/materials/LinearInterpolationMaterial.C#L54-L58

There's nothing wrong with this code other than any issue you might have with using Exception handling for "control flow" design issues. Here's we are calling an external utility and catching an exception so that we can provide a more context aware error message.

**Comment by permcody** [comment]
> While we sort out what we should do to fix the try/catch crap. Maybe we can push the hash forward a bit more? I wouldn't mind picking up the latest Nemesis stuff to start playing with:
> [libMesh/libmesh#2338](https://github.com/libMesh/libmesh/pull/2338)

@lindsayad - want to advance your libMesh hash forward? 

Jason is busy creating a new Mac test machine with Conda environment on it instead of the existing package. Since we really hoping to move to Conda in the near term, this might be the way to go. I'm hoping it just magically fixes the exception problem so I don't have to do something ugly to perfectly valid C++ code.

**Comment by lindsayad** [comment]
K advanced

**Comment by lindsayad** [comment]
What is going on with CIVET?!

**Comment by permcody** [comment]
We have a rogue build box! This is a problem I've thought about several times. Super hard to solve but also very important:
https://github.com/idaholab/civet/issues/466 

We need an admin to kill mobuild58. It's totally gone rogue as in we can't remote into it to kill it, but it's still alive enough to pick up jobs and fail them - haha!

**Comment by lindsayad** [comment]
@permcody can you share your stand-alone code?

**Comment by permcody** [comment]
The exception is not caught on our Mac boxes.

**Comment by lindsayad** [comment]
This is ridiculous that this Mac bull*$@% is holding up so much stuff. I have so many PRs that are waiting on this. There just has to be some mismatch between the type info of the run-type exception and the type info of the exceptions stored in the `.gcc_except_table` generated at compile time for catch landing points. But I don't know how in the world to debug this since obviously the system libraries do not have debug information.

**Comment by permcody** [comment]
> This is ridiculous that this Mac bull*$@% is holding up so much stuff.

OK - I know you are super frustrated with it, but you guys did make a lot of progress on this this weekend. It appears that the exception handling stuff is completely fixed with Jason's latest build! The issue now is all related to a bunch of Python crap, which I'm totally onboard with skipping if we need to. Jason knows what's going on there and he's working on another package to get that fixed up too. Perhaps we'll have a working package by tomorrow (12/10)? If not, I'm open to temporarily skipping all of the remaining broken MOOSEDocs tests temporarily to get things moving again.

Bottom Line: We should be able to move forward tomorrow regardless of what happens with the package.

**Comment by permcody** [comment]
BTW - I like the fact that your messages is "edited" and also contains grawlixes. I wonder what the original looked like - hehe.

**Comment by roystgnr** [comment]
> It appears that the exception handling stuff is completely fixed with Jason's latest build!

Whoa - I hadn't noticed that.  Is this because "Conda ... just magically fixes the exception problem", or did we actually pin down the underlying bug?  No complaints either way, just if someone *does* understand the bug I'd love to find out the details.

**Comment by lindsayad** [comment]
What did @milljm actually change in this newest package? Because in the old package, if I ran `otool -hV` on every single library that `moose-test_dbg` linked against, every single one was listed as `TWOLEVEL`

**Comment by lindsayad** [comment]
@roystgnr I still don't think we have a real idea of what is causing this. If we have it fixed, (which it looks like @milljm does), I think it's mostly through Monte Carlo

**Comment by lindsayad** [comment]
It would be so cool if Mac offered debug versions of their system libraries that we could link against... I get annoyed at forums that say "why would you want to look at system code? System code should be a block box. Only worry about your application code". 😠 I would be so happy if I could step through this in the debugger:
```c
            while (true)
            {
                const uint8_t* actionRecord = action;
                int64_t ttypeIndex = readSLEB128(&action);
                if (ttypeIndex > 0)
                {
                    // Found a catch, does it actually catch?
                    // First check for catch (...)
                    const __shim_type_info* catchType =
                        get_shim_type_info(static_cast<uint64_t>(ttypeIndex),
                                           classInfo, ttypeEncoding,
                                           native_exception, unwind_exception);
                    if (catchType == 0)
                    {
                        // Found catch (...) catches everything, including foreign exceptions
                        // If this is a type 1 search save state and return _URC_HANDLER_FOUND
                        // If this is a type 2 search save state and return _URC_HANDLER_FOUND
                        // If this is a type 3 search !_UA_FORCE_UNWIND, we should have found this in phase 1!
                        // If this is a type 3 search _UA_FORCE_UNWIND, ignore handler and continue scan
                        if ((actions & _UA_SEARCH_PHASE) || (actions & _UA_HANDLER_FRAME))
                        {
                            // Save state and return _URC_HANDLER_FOUND
                            results.ttypeIndex = ttypeIndex;
                            results.actionRecord = actionRecord;
                            results.landingPad = landingPad;
                            results.adjustedPtr = get_thrown_object_ptr(unwind_exception);
                            results.reason = _URC_HANDLER_FOUND;
                            return;
                        }
                        else if (!(actions & _UA_FORCE_UNWIND))
                        {
                            // It looks like the exception table has changed
                            //    on us.  Likely stack corruption!
                            call_terminate(native_exception, unwind_exception);
                        }
                    }
                    // Else this is a catch (T) clause and will never
                    //    catch a foreign exception
                    else if (native_exception)
                    {
                        __cxa_exception* exception_header = (__cxa_exception*)(unwind_exception+1) - 1;
                        void* adjustedPtr = get_thrown_object_ptr(unwind_exception);
                        const __shim_type_info* excpType =
                            static_cast<const __shim_type_info*>(exception_header->exceptionType);
                        if (adjustedPtr == 0 || excpType == 0)
                        {
                            // Something very bad happened
                            call_terminate(native_exception, unwind_exception);
                        }
                        if (catchType->can_catch(excpType, adjustedPtr))
                        {
                            // Found a matching handler
                            // If this is a type 1 search save state and return _URC_HANDLER_FOUND
                            // If this is a type 3 search and !_UA_FORCE_UNWIND, we should have found this in phase 1!
                            // If this is a type 3 search and _UA_FORCE_UNWIND, ignore handler and continue scan
                            if (actions & _UA_SEARCH_PHASE)
                            {
                                // Save state and return _URC_HANDLER_FOUND
                                results.ttypeIndex = ttypeIndex;
                                results.actionRecord = actionRecord;
                                results.landingPad = landingPad;
                                results.adjustedPtr = adjustedPtr;
                                results.reason = _URC_HANDLER_FOUND;
                                return;
                            }
```

**Comment by lindsayad** [comment]
Maybe I will build my own `libc++` and `libc++abi` with debugging symbols and then set `CPPFLAGS and `LDFLAGS`...

**Comment by permcody** [comment]
@lindsayad - you are venturing into knowledge that only the elders hold...

@roystgnr - Jason said that since all of these problems seemed to stem from MPICH, he decided to build his own with the two-level-namespace option, but I'm still not convinced. I swore his old package also had that option so I need to ask him what changed there. He's not using Conda to fix this as the problem also occurred in Conda. I tried to bisect libMesh but unfortunately there are a few commits in master that didn't build related to the "passmess" refactor. I can say that the commit right before the first passmess commit still built and tested OK. It was after passmess (or TIMPI) not sure which, where it magically started to fail. As Alex said, it might all be some sort of Monte Carlo thing but it is a problem on Macs that we do have to solve unless we are ready to abandon our Macs!

**Comment by lindsayad** [comment]
> Jason said that since all of these problems seemed to stem from MPICH, he decided to build his own with the two-level-namespace option, but I'm still not convinced. I swore his old package also had that option so I need to ask him what changed there.

I'm not convinced either although @milljm is a master at his craft. As I said [here](https://github.com/idaholab/moose/pull/14464#issuecomment-563485493) I'm using a moose package on Catalina that demonstrates the complete lack of exception catching, but every library that `moose-test_db` links against register as being `TWOLEVEL`

**Comment by lindsayad** [comment]
@roystgnr I'm pretty much positive I know what the issue is. libmesh moves `/usr/lib` to the end of the link line but I think timpi isn't doing that and per `libtimpi_dbg.la` `dependency_libs=-L/usr/lib` ...so we are linking to the system libc++ instead of our package provided libc++

**Comment by permcody** [comment]
Very nice detective work. I bet that’s it!

**Comment by lindsayad** [comment]
Oh yeaa...see below!!! @milljm my suggestion is that we add to our mpi module: `export MPIHOME=<wherever_mpi_home_is>`. This way timpi will correctly pick up the MPI location and not introduce bad flags like `-I/usr/include` and `-L/usr/lib` which it does by default when it doesn't detect the MPI location (all it currently tries is `MPIHOME` in the environment as well as the `--with-mpi=` configure flag).

Another solution would be to get an autotools expert like @jwpeterson (or @roystgnr ?) to change the `mpi.m4` script in TIMPI (and probably in libMesh) to be able to smartly detect the mpi directory from the path of `which mpicc` or something like that if that program exists. I know I can do `AC_PATH_PROG` but then I dunno how to strip off `/bin/mpicc` except by calling shell commands (`syscmd` and `esyscmd`) but I feel like there has to be a simpler way.
```
(base) lindad@Alexanders-MacBook-Pro:~/projects/moose/test(libmesh-update)$ otool -L moose_test-dbg 
moose_test-dbg:
	/Users/lindad/projects/moose/test/lib/libmoose_test-dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/lindad/projects/moose/framework/libmoose-dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/lindad/projects/moose/framework/contrib/pcre/libpcre-dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/lindad/projects/moose/framework/contrib/hit/libhit-dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/lindad/projects/moose/libmesh/installed-module-mpich/lib/libmesh_dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/lindad/projects/moose/libmesh/installed-module-mpich/lib/libnetcdf.13.dylib (compatibility version 15.0.0, current version 15.1.0)
	/Users/lindad/projects/moose/libmesh/installed-module-mpich/lib/libtimpi_dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-opt/lib/libslepc.3.11.dylib (compatibility version 3.11.0, current version 3.11.0)
	/opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-opt/lib/libpetsc.3.11.dylib (compatibility version 3.11.0, current version 3.11.4)
	/opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-opt/lib/libsuperlu_dist.6.dylib (compatibility version 6.0.0, current version 6.1.1)
	/opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-opt/lib/libparmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/moose/petsc-3.11.4/mpich-3.3_clang-9.0.0-opt/lib/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/X11/lib/libX11.6.dylib (compatibility version 10.0.0, current version 10.0.0)
	/opt/moose/mpich-3.3/clang-9.0.0/lib/libmpifort.12.dylib (compatibility version 14.0.0, current version 14.6.0)
	/opt/moose/gcc-9.2.0/lib/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	/opt/moose/gcc-9.2.0/lib/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/opt/moose/llvm-9.0.0/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/opt/moose/mpich-3.3/clang-9.0.0/lib/libmpicxx.12.dylib (compatibility version 14.0.0, current version 14.6.0)
	/opt/moose/mpich-3.3/clang-9.0.0/lib/libmpi.12.dylib (compatibility version 14.0.0, current version 14.6.0)
	/opt/moose/mpich-3.3/clang-9.0.0/lib/libpmpi.12.dylib (compatibility version 14.0.0, current version 14.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/opt/moose/gcc-9.2.0/lib/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.27.0)
	/opt/moose/gperftools-2.7/lib/libtcmalloc_and_profiler.4.dylib (compatibility version 10.0.0, current version 10.3.0)
	/System/Library/Frameworks/OpenCL.framework/Versions/A/OpenCL (compatibility version 1.0.0, current version 1.0.0)
	/opt/moose/llvm-9.0.0/lib/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
(base) lindad@Alexanders-MacBook-Pro:~/projects/moose/test(libmesh-update)$ cd tests/misc/check_error/
(base) lindad@Alexanders-MacBook-Pro:~/projects/moose/test/tests/misc/check_error(libmesh-update)$ ../../../moose_test-dbg -i linear_interp_not_increasing.i 



*** ERROR ***
In LinearInterpolationMaterial linear_interp: x-values are not strictly increasing: x[1]: 0.2 x[2]: 0.2
```

**Comment by roystgnr** [comment]
> to change the mpi.m4 script in TIMPI (and probably in libMesh)

At first guess I'd assume I got them out of sync somehow when trying to break things apart (perhaps when I was trying to be "clean" about it before I reintroduced the PETSc-scraping option?)  If it's a bug in the libMesh script too, how were we not seeing the same consequences before the TIMPI refactor?

**Comment by lindsayad** [comment]
> If it's a bug in the libMesh script too, how were we not seeing the same consequences before the TIMPI refactor?

This isn't a bug in my opinion. The link line is formed with the help of dependency libs from `libcontrib*.la`. Now that TIMPI is in contrib, it contributes to the `libcontrib` dependency libs. If TIMPI isn't given a `MPIHOME` it throws in `/usr/lib` to the dependency libs and then that gets into the libmesh link line. If `/usr/lib` was getting added to the link line from `libmesh_optional_LIBS` that would be ok because we sort and move system libraries to the end in the `libmesh_optional_packages.m4` script. But dependencies from `libcontrib` are put at the beginning of the link line, e.g. see `Makefile.am`:
<pre>
if LIBMESH_DBG_MODE
  lib_LTLIBRARIES        += libmesh_dbg.la
  libmesh_dbg_la_SOURCES  = $(libmesh_SOURCES)
  libmesh_dbg_la_CPPFLAGS = $(CPPFLAGS_DBG) $(AM_CPPFLAGS)
  libmesh_dbg_la_CXXFLAGS = $(CXXFLAGS_DBG)
  libmesh_dbg_la_CFLAGS   = $(CFLAGS_DBG)
  <b>libmesh_dbg_la_LIBADD   = contrib/libcontrib_dbg.la $(LIBS)</b>
endif
</pre>

**Comment by jwpeterson** [comment]
> If TIMPI isn't given a MPIHOME it throws in /usr/lib to the dependency libs

IMO this is a bug. We've all spent lots of time trying to prevent rogue MPIs from being linked in by accident (thus the sorting and moving of system libs that you discovered) so if it's now possible for TIMPI to do this then it should be fixed.

**Comment by lindsayad** [comment]
> IMO this is a bug. We've all spent lots of time trying to prevent rogue MPIs from being linked in by accident (thus the sorting and moving of system libs that you discovered) so if it's now possible for TIMPI to do this then it should be fixed.

Well I guess that's a valid point. Especially because there's actually no mpi stuff at all in the system.

**Comment by jwpeterson** [comment]
Yeah, having -L/usr/lib anywhere at the beginning of a link line on Ubuntu systems is a recipe for problems. Aside from MPI there can be libmetis and libpetsc installed from `apt-get` that can be pulled in by accident this way...

**Comment by roystgnr** [comment]
> Yeah, having -L/usr/lib anywhere at the beginning of a link line on Ubuntu systems is a recipe for problems.

On any systems.  There's more than a few people in situations where their OS includes /usr/lib/libFooBar.0.ancient, and they want to use /opt/modules/foobar/1.new/lib/libFooBar.1.new instead, and their sysadmin won't remove the ancient version because other users depend on it.  This is definitely a bug.

In my defense, though, I'm pretty confident that I didn't actually *add* that behavior, I just somehow didn't replicate the libMesh detection flow perfectly and so somehow started us newly *triggering* that behavior.

**Comment by lindsayad** [comment]
> In my defense, though, I'm pretty confident that I didn't actually _add_ that behavior, I just somehow didn't replicate the libMesh detection flow perfectly and so somehow started us newly _triggering_ that behavior.

You're definitely correct. Your logic at the top of the TIMPI m4 is the exact same as that in libMesh's.

**Comment by roystgnr** [comment]
> Your logic at the top of the TIMPI MPI m4 is the exact same as that in libMesh's.

Is it possible that somehow libMesh petsc.m4 is detecting PETSc's MPI configuration but TIMPI petsc_mpi.m4 isn't?

Could you send me both libMesh and TIMPI config.logs from a system that fails without MPIHOME set?

**Comment by lindsayad** [comment]
> Could you send me both libMesh and TIMPI config.logs from a system that fails without MPIHOME set?

I'll get this info to you, but if you look in libMesh's m4's, we only ever call `ACX_MPI` if we can't snoop MPI from PETSc, the PETSc configuration failed, or the user requested `--disable-petsc`. This all happens in `petsc.m4`. Conversely, in TIMPI you only run the PETSc m4 script if the `ACX_MPI` detection of MPI fails.

**Comment by lindsayad** [comment]
@roystgnr:

[libmesh config](https://github.com/idaholab/moose/files/3947496/config.log)
[timpi config](https://github.com/idaholab/moose/files/3947514/config.log)

**Comment by roystgnr** [comment]
> Conversely, in TIMPI you only run the PETSc m4 script if the ACX_MPI detection of MPI fails.

Damn, of course that's it.  I'd been thinking of that as a patch-up to the case where ACX_MPI failed; it didn't occur to me that there would be cases where ACX_MPI would "pass" but with an inferior configuration.

**Comment by permcody** [comment]
OK - so the mac failure is right out of the twilight zone. I used the linear partitioner and reran the tests (regolded) yet they still appear flipped on the test machines. I just pulled off this last commit to see if we even need it now since I regolded because they were flipped in the first place. I don't understand why this test is all of a sudden unstable. It's restricted to Mac and it's a simple 4 element test. 

If we can't get it to behave, I'm going to switch it to "replicated" instead of "distributed" since we don't need to test it with distributed. It's just a test of nemesis and I don't really care to test any behaviors of parmetis as a side effect.

**Comment by fdkong** [comment]
> OK - so the mac failure is right out of the twilight zone. I used the linear partitioner and reran the tests (regolded) yet they still appear flipped on the test machines. I just pulled off this last commit to see if we even need it now since I regolded because they were flipped in the first place. I don't understand why this test is all of a sudden unstable. It's restricted to Mac and it's a simple 4 element test.
> 
> If we can't get it to behave, I'm going to switch it to "replicated" instead of "distributed" since we don't need to test it with distributed. It's just a test of nemesis and I don't really care to test any behaviors of parmetis as a side effect.

Linear partitioner should work. I do not quite understand why the linear partitioner did not work here

**Comment by permcody** [comment]
> Linear partitioner should work. I do not quite understand why the linear partitioner did not work here.

I agree, but I also don't understand why this is breaking now. I verified that the actual combined solution is still the same between the old and the new. This really does boil down to a partitioning issue. I don't understand why this is breaking now. We didn't change PETSc or anything.

**Comment by lindsayad** [comment]
Oh my god, is it going to work?!

**Comment by permcody** [comment]
Here we go!!!

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
@snschune or @dschwen - It looks like Magpie is going to need a patch related to TIMPI (the new C++ MPI Wrapper library that is being split out). Do either of you have time to look at the failure?

**Comment by dschwen** [comment]
I can deal with that. Any examples on what to do?

**Comment by permcody** [comment]
Since Magpie was compiling before, my guess is that somehow you are just missing an include now (due to the refactor of parallel -> TIMPI). Look at the new TIMPI stuff and see if there's a new header or something that has the missing specialization that Magpie is complaining about.

**Comment by roystgnr** [comment]
I haven't seen the failure message, but there's a few possibilities.  The most likely IMHO might be a forward declaration that needs a change of namespace; using declarations provide backward compatibility for most uses of the old namespace but not for forward declarations, except for Communicator itself which is so commonly forward-declared that I made a whole shim class for it.  The using declarations don't work for packed range specializations, either, and the workarounds are even worse, so at some point we need to do a wholesale libMesh::Parallel->TIMPI upgrade on the Packing class.

**Comment by permcody** [comment]
There's a whole pile of these:
```
/opt/civet/build_0/magpie/build/header_symlinks/RadialGreensConvolution.h:176:7: error: specialization of 'template<class T> class TIMPI::StandardType' in different namespace [-fpermissive]
  176 | class StandardType<RadialGreensConvolution::QPData> : public DataType
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/civet/build_0/moose/scripts/../libmesh/installed/include/libmesh/parallel.h:45,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/DataIO.h:24,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/MaterialProperty.h:16,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/MaterialData.h:12,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/BlockRestrictable.h:15,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/KernelBase.h:13,
                 from /opt/civet/build_0/moose/framework/build/header_symlinks/Kernel.h:12,
                 from /opt/civet/build_0/magpie/build/header_symlinks/CoupledDefectAnnihilation.h:11,
                 from /opt/civet/build_0/magpie/src/kernels/CoupledDefectAnnihilation.C:9,
                 from /opt/civet/build_0/magpie/build/unity_src/kernels_Unity.C:2:
/opt/civet/build_0/moose/scripts/../libmesh/installed/include/timpi/standard_type.h:60:7: note:   from definition of 'template<class T> class TIMPI::StandardType'
   60 | class StandardType : public DataType
      |       ^~~~~~~~~~~~
In file included from /opt/civet/build_0/magpie/build/header_symlinks/RadialGreensSource.h:12,
                 from /opt/civet/build_0/magpie/src/kernels/RadialGreensSource.C:9,
                 from /opt/civet/build_0/magpie/build/unity_src/kernels_Unity.C:6:
```

**Comment by roystgnr** [comment]
Ah!  That's right; Packing wasn't the only class where user specializations were expected.  In hindsight I should have kept StandardType defined in (rather than just using-declaration'ed in) libMesh::Parallel too, but at this point it'd probably be less work to change the namespaces in the application.

**Comment by dschwen** [comment]
Fixes are up here: https://github.com/idaholab/magpie/pull/408

Basically only namespace changes.

**Comment by dschwen** [comment]
I haven't seen a whole lot of uses of StandardType specializations. It seems to me that for linearly laid out data this is a more efficient solution than packing.

**Comment by roystgnr** [comment]
Oh, definitely!  The packed_range code was intended to be *solely* for variable-length, heap allocated, etc. data.  For arrays of fixed-structure data then it's just going to add unnecessary buffering.

**Comment by lindsayad** [comment]
@roystgnr We may perform another libmesh update this week because of the awesomeness of libmesh/libmesh#2377. One thing we want to do is fixup the TIMPI MPI detection. I'm perfectly fine handling that unless you're already doing it.

**Comment by roystgnr** [comment]
Sounds reasonable.  I'd prefer you handle it, since the current bugginess seems to be more blatant on your systems - I *think* I have a new computer that's showing related build issues but I'm only 50% confident on that.
